spirv_msl.cpp

```diff
@@ -19389,6 +19394,70 @@ void CompilerMSL::emit_mesh_tasks(SPIRBlock &block)
 	statement("return;");
 }
 
+void CompilerMSL::emit_workgroup_initialization(const SPIRVariable &var)
+{
+	auto &type = get_variable_data_type(var);
+
+	// We need to use gl_LocalInvocationIndex, make sure it's available
+	ensure_builtin(StorageClassInput, BuiltInLocalInvocationIndex);
+
+	begin_scope();
+
+	if (type.array.empty() && type.member_types.empty()) {
+		// For simple shared variables, we just initialize it in thread 0 of the block
+		// We use short to represent bool for threadgroup variable to workaround compiler bug,
+		// so we do a temporary fixup here. Alas. (see the type_to_glsl method)
+		bool is_boolean = type.basetype == SPIRType::Boolean;
+		if (is_boolean) {
+			type.basetype = SPIRType::Short;
+		}
+		statement("if (gl_LocalInvocationIndex == 0)");
+		begin_scope();
+		statement(join(to_name(var.self), " = ", to_initializer_expression(var), ";"));
+		end_scope();
+		if (is_boolean) {
+			type.basetype = SPIRType::Boolean;
+		}
+	} else {
+		// Otherwise, we use a loop to cooperatively initialize the memory within the group
+
+		// First, we define a few variable names;
+		string var_name = to_name(var.self);
```

Comment:
This really needs test output to be included.

Reply:
Two tests added to reflect the two cases of this function.